opentype-shaping-gurmukhi.md

```diff
@@ -1029,12 +1031,63 @@ so that glyphs of the same ordering category remain in the same
 relative position with respect to each other.
 
 
+#### 2.10: Flag sequences for possible feature applications ####
+
+With the inital reordering complete, those glyphs in the syllable that
+may have GSUB or GPOS features applied in stages 3, 5, and 6 should be
+flagged for each potential feature. 
+
+This flagging is preliminary; the set of potential features varies
+between different scripts and which features are supported varies
+between fonts. It is also possible that the application of
+one feature on a glyph sequence will perform a substitution that makes
+a later feature no longer applicable to the updated sequence.
+
+Consequently, the flagging must be completed before shaping procedes
+to the stages during which features are applied.
+
+Some shaping features, such as `locl`, can potentially apply to any
+glyphs. Therefore it is not necessary to maintain a separate flag for
+these features in the bitmask (or other data structure) used to track
+the flags -- although shaping engines may do so if desired.
+
+The sequences to flag are summarized in the list below; a full
+description of each feature's function and interpretation is provided
+in GSUB and GPOS application stages that follow.
+
+  - `nukt` should match "_Consonant_,Nukta" sequences
+  - `akhn` should match "Ka,Halant,Ssa" and "Ja,Halant,Nya"
+  - `rphf` should match initial "Ra,Halant" sequences but _not_ match
+            initial "Ra,Halant,ZWJ" sequences
+  - `blwf` should match "Halant,Ra", "Halant,Ha", and "Halant,Va" in
+            post-base positions and "Ra,Halant", "Ha,Halant", and
+            "Va,Halant" in non-initial pre-base positions
+  - `half` should match "_Consonant_,Halant" in pre-base position but
+           _not_ match "Ra,Halant" sequences flagged for `rphf` and
+           _not_ match "_Consonant_,Halant,ZWNJ,_Consonant_" sequences
+  - `pstf` should match initial "Halant,Ya" in post-base position
+  - `vatu` should match "_Consonant_,Halant,Ra",
+           "_Consonant_,Halant,Ha", and "_Consonant_,Halant,Va"
+  - `cjct` should match "_Consonant_,Halant,_Consonant_" but _not_
+            match "_Consonant_,Halant,ZWJ,_Consonant_" or
+            "_Consonant_,Halant,ZWNJ,_Consonant_"
+  - `init` should match word-initial letters
+  - `haln` should match word-final "_Consonant_,Halant" sequences
```

Comment:
Yes, it's stage 5. I honestly put it in here just for adjacency with `init`, because they are both positionally contextual. I would think (as a non–shaping-engine-writer) that if a test for word-iniitalness is required for `init` (as Allsorts [does](https://github.com/yeslogic/allsorts/blob/d5525fb0d1defcb981e04fdfab2e3c1bdc4a2499/src/indic.rs#L2340)) then the advice would also be to handle `haln` the same way. 

That may largely be a moot point considering that doing `haln` in GSUB seems really uncommon (in fact I'm not certain I have an example of it), but if that's the right way to handle the feature it would make sense to discuss it with `init`.

Reply:
> I honestly put it in here just for adjacency with `init`, because they are both positionally contextual.

That makes sense.